### PR TITLE
[ENH] Signature Moments transformer 

### LINKED ISSUE #7316: [ENH] discrete signature feature transformer
It would be great to have a simple, no-bells-and-whistles signature feature transformer implemented in `sktime`.

Current implementations suffer either from overengineering (and/or unclear, unstable interfaces), or dependency on abandoned packages relying on lower python versions, see discussion in https://github.com/sktime/sktime/issues/7255

However, the idea behind signatures is quite simple, they are time-ordered generalizations of moments for multivariate time series, I describe it in the following for a bivariate time series.

Let $(x_{1,1}, x_{2,1}) , \dots, (x_{1,T}, x_{2,T})$ be a bivariate time series.

Elements of the "signature" are indexed by strings made up of the literals 1, and 2.
I list the signature elements for the strings up to length 3, the general pattern should follow.

1: the mean of the ${x_{1, i}}$, over all $i$.

2: the mean of the ${x_{2, i}}$, over all $i$.

11: the mean of all products ${x_{1, i}\cdot x_{1, j}}$, over all $i<j$.

12: the mean of all products ${x_{1, i}\cdot x_{2, j}}$, over all $i<j$.

$\ell m$ : the mean of all products ${x_{\ell, i}\cdot x_{m, j}}$, over all $i<j$.

$\ell m n$: the mean of all products ${x_{\ell, i}\cdot x_{m, j}\cdot x_{n, k}}$, over all $i<j<k$.

And so on. Generalisation to longer strings, or more dimensions should be clear.

Due to combinatorial explosion, it typically does not make sense to compute higher than 3, or 2 for high-dimensional time series.

The transformer should be parameterized in terms of an integer `degree`, that is the length of strings to include. Often, the index of the time series is included as a dimension, so for univariate time series, one gets two dimension together with the time index. This should be a boolean option, e.g., `use_index`.

The transformer is "series-to-primitives" type, the supersimple extension template should be sufficient to follow: https://github.com/sktime/sktime/blob/main/extension_templates/transformer_supersimple.py

The `scitypes:transform-output` tag should change to `"Primitives".

### PR DESCRIPTION
**Pull request by VectorNd**
Towards #7316 

Implemented signature feature transformer for multivariate time series  having degree upto 3 and after this degree not implemented due to combinatorial explosion 

- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice, great!

Some change requests, blocking:

* the existing transformation should not be changed. Create a new one.
* This PR also includes the one with the data loaders, I believe this is a branching error. Since we squash PR, you can revert manually by using git, or copy-pasting the `datasets` folder from `main` and committing. Neither will impact the commit history on `main` once merged.
* the docstring should explain that `degree` can only be up to 3.
* the docstring should explain the math, to avoid misunderstandings. The implemented features are not exactly the signature, you get (a discrete approximation of) the signature when you pipeline with first differences. How about calling the transformer `SignatureMoments` or similar? Happy to help with this part if not clear.

Non-blocking:

* it would be useful to have this implemented for arbitrary degree, but I recognize that may be more difficult.
* it would be nice to use `numpy` broadcasting instead of list comprehensions for efficiency. You could use matrix tricks etc, I will paste some code I have.

**Comment by fkiraly** [comment]
For the last point, here is some `numpy` code, generated in collaboration with a chatbot, to sum over upper triangular parts of an nd array. This is potentially useful to replace the list comprehension.

I note this will probably use more memory but be faster than the list comprehension, this may be mitigated by using sparse matrices.

I have not benchmarked or tested this extensively, I am just pasting here as it may help.


```python
import numpy as np

def sum_upper_simplices_nd(arr):
    n = arr.shape[0]  # Number of elements along the first dimension
    dimensions = arr.ndim  # Number of dimensions of the array

    # Create an array of indices
    indices = np.arange(n)

    # Create a meshgrid for indices
    mesh = np.meshgrid(*([indices] * dimensions), indexing='ij')  # Creates a grid of indices

    # Stack the meshgrid into an array
    index_array = np.stack(mesh)  # Shape (dimensions, n, n, ..., n)

    # Use broadcasting to create a mask for upper simplices
    # This compares all combinations of indices and checks if they are in strictly increasing order
    mask = np.all(np.diff(index_array, axis=0) > 0, axis=0)  # Shape will be (n, n, ..., n)

    # Sum the elements where the mask is True
    total_sum = np.sum(arr[mask])

    return total_sum
```

**Comment by fkiraly** [comment]
FYI @moschnetzsch

**Comment by fkiraly** [comment]
Also, one thing I notice, maybe it should be "smaller equal" in the inequalities, that is, $i \le j \le k$. Perhaps we can make this an parameter, whether we use strict or inequality.

In the broadcasting code, this would correspond to a change `mask = np.all(np.diff(index_array, axis=0) >= 0, axis=0)`.

**Comment by VectorNd** [comment]
@fkiraly  i have made some changes  . Please review it and let me know if i need to do anything

File: sktime/datasets/_single_problem_loaders.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
     "jasonlines",
     "achieveordie",
     "ciaran-g",
-    "jonathanbechtel",
+    "jonathanbechtel"
```
Comment (fkiraly) [inline sktime/datasets/_single_problem_loaders.py:16]: these commas should remain, in accordance with code formatting requirements. Is your local linter setting automatically removing them? See here how to set up linting locally so it agrees with the conventions:
https://www.sktime.net/en/stable/developer_guide/coding_standards.html#setting-up-local-code-quality-checks

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great, I think this is almost ready to go!

* you seem to have unusual linting settings, which are also applied to other files - could you revert them and use the `sktime` code formatting settings? See above.
* in `_transform`, you loop over `instance_idx` - so there is no efficiency gain from using a panel data format. I would suggest you use `X_inner_mtype : "pd.DataFrame"`, which returns a single series already in 2D `pd.DataFrame` format, (timepoints, variables). You need to return a single-row `pd.DataFrame`
* `get_test_params` should return at least two parameter settings, an optimally cover different parameters of the `__init__`

File: sktime/datasets/_single_problem_loaders.py (hunk 1 above)
Comment (fkiraly) [reply sktime/datasets/_single_problem_loaders.py]: fixed this now

**Review by fkiraly — APPROVED** [review]
I now fixed the formatting problems and outstanding issues, so @moschnetzsch can have this in the next release. Also added a normalization option.

**Review by fkiraly — APPROVED** [review]
Also fixed a side effects via list bug.

@VectorNd, FYI, lists are mutable and are not automatically reset to their initial state. So, if you are using a list multiple times, you need to be aware that it will change by reference rather than just by return.

Good prompt for chatbots: "explain the most common type of problems that can occur from mutating lists in python".

**Comment by fkiraly** [comment]
FYI @moschnetzsch, would appreciate a review.

**Pull request merged**